PSP for JSON-RPC (#6)
w3f/PSPs: PSPs/drafts/psp-6.md

## 1.2. Motivation

This standard should serve as  common ground for API endpoints in order for external applications to be able to reuse existing functionality on various implementations of the Polkadot node. Avoiding disparity among node implementations will simplify their development, as well as allow applications which use different node implementations to switch seamlessly between them.

## 1.3. Specification

[JSON-RPC](https://www.jsonrpc.org/specification) is a stateless, light-weight remote procedure call (RPC) protocol. Primarily this specification defines several data structures and the rules around their processing. It is transport agnostic in that the concepts can be used within the same process, over sockets, over http, or in many various message passing environments. It uses [JSON](https://www.json.org/json-en.html) ([RFC 4627](https://www.ietf.org/rfc/rfc4627.txt)) as data format.

All parameters in the specified JSON-RPC methods are **REQUIRED**, unless explicitly mentioned otherwise (**OPTIONAL**). Return values which are OPTIONAL indicate the possibility of `null` being returned.

> Additionally, the JSON-RPC API should be made available over the [WebSocket Protocol](https://tools.ietf.org/html/rfc6455).

Author: Done. Just note that there is no official websocket `pubsub` standard.

| API / Network | Localhost                 | All interfaces             |
| ------------- | ------------------------- | -------------------------- |
| Safe API      | `--local-{rpc/ws}`        | `--public-{rpc/ws}`        |
| Unsafe API    | `--local-unsafe-{rpc/ws}` | `--public-unsafe-{rpc/ws}` |

**Warning**: Enabling `--public-unsafe-{rpc|ws}` is dangerous and should be avoided.

### 1.3.2. Errors

Each namespace specifies RPC error code ranges of possible error types that can be returned. Known error types are defined in this document and some APIs may return undefined error messages. Defined error messages are subject to expansion and changes.

### 1.3.3. Common types

This document references commonly used types, represented in capital letters, as defined in the subsections below.

#### 1.3.3.1. STRING

A sequence of characters: `"value"`

#### 1.3.3.2. U32 / U64

Unsigned integers.

* `U32` - A 32-byte unsigned integer.
* `U64` - A 64-byte unsigned integer.

#### 1.3.3.3. BOOL

A boolean type, which can either be `true` or `false`.

#### 1.3.3.4. MAP

> A key/value object.

Author: Yes, I added this section:

> A key/value object, where keys are `STRING`s and values are arbitrary types. Each API defines its own value types.

```json
{
    "key": "value1",
    "key2": "value2"
}
```

#### 1.3.3.5. ARRAY

An object containing a varying amount of items.

```json
[
    "item1",
    "item2"
]
```

#### 1.3.3.6. HASH

A string which represents a 32-byte, hexadecimal-encoded Blake2 hash. Contains a `0x` prefix to indicate the hexadecimal encoding.

> #### 1.3.3.7. HEX
> 
> A string of varying size representing hexadecimal-encoded data. Contains a `0x` prefix to indicate the hexadecimal encoding.

Reviewer: Would it be better to put this before `HASH` since all hashes are hex? Also, consider explicitly enumerating the valid characters, e.g. `[a-fA-F0-9]`

```suggestion
#### 1.3.3.7. HEX

A variable-size string that represents hexadecimal-encoded data. Must begin with the characters `0x` to indicate a hexadecimal encoding, e.g. `0x1337`.
```

Author: > Would it be better to put this before HASH since all hashes are hex?

Done!

> Also, consider explicitly enumerating the valid characters, e.g. [a-fA-F0-9]

Given that it explicitly states "hexadecimal-encoded", I guess this is somewhat redundant. But I added it anyway.

    - [1.13.2. engine_createBlock](#1132-engine_createblock)
    - [1.13.3. engine_finalizeBlock](#1133-engine_finalizeblock)
  - [1.14. Payment](#114-payment)
    - [1.14.1. Payment Errors](#1141-payment-errors)
    - [1.14.2. payment_queryInfo](#1142-payment_queryinfo)
  - [1.15. Contracts](#115-contracts)
    - [1.15.1. Contracts Errors](#1151-contracts-errors)
    - [1.15.2. contracts_call](#1152-contracts_call)
    - [1.15.3. contracts_getStorage](#1153-contracts_getstorage)
      - [1.15.3.1. Parameter](#11531-parameter)
      - [1.15.3.2. Response](#11532-response)
    - [1.15.4. contracts_rentProjection](#1154-contracts_rentprojection)
  - [1.16. BABE](#116-babe)
    - [1.16.1. BABE Errors](#1161-babe-errors)
    - [1.16.2. babe_epochAuthorship](#1162-babe_epochauthorship)
      - [1.16.2.1. Parameter](#11621-parameter)
      - [1.16.2.2. Response](#11622-response)
  - [1.17. Sync](#117-sync)
    - [1.17.1. sync_state_genSyncSpec](#1171-sync_state_gensyncspec)
      - [1.17.1.1. Parameter](#11711-parameter)
      - [1.17.1.2. Response](#11712-response)
      - [1.17.1.3. Example](#11713-example)
  - [1.18. Copyright](#118-copyright)

## 1.5. System

> System RPC API.

Reviewer: Can this description be improved a bit? It doesn't feel like it adds much at this time. Maybe something along the lines of this?
```suggestion
RPC endpoints that expose system metadata or can be used for administration of the blockchain network.
```

### 1.5.1. System Errors

RPC error codes are in the `2000` - `2999` range. No known error types are specified.

> ### 1.5.2. system_name

Reviewer: For names like this consider formatting as code...

```suggestion
### 1.5.2. `system_name`
```

Get the node's implementation name.

#### 1.5.2.1. Parameter

None.

#### 1.5.2.2. Response

`STRING` - The node's name.

#### 1.5.2.3. Example

Request:

```bash
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "system_name", "params":[]}' http://localhost:9933
```

Response:

```json
{
    "jsonrpc": "2.0",
    "result": "parity-polkadot",
    "id": 1
}
```

### 1.5.3. system_version

> Get the node implementation's version. Should be a semver string.

Reviewer: Watch out for jargon like "semver"...if you use it, link out to a resource where the reader can learn more about it.

* `HASH` - The block hash indicating the ending of the range.
* `U64` - (OPTIONAL) The Authority Set ID.

#### 1.7.2.2. Response

* `HEX` - (OPTIONAL) The SCALE encoded array of finality proofs.

#### 1.7.2.3. Example

Request:

```bash
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "grandpa_proveFinality", "params":["0xd8c6bfd22432e1fdf2fbe4e6f32fb650b7bb5a9c4c992d53129bd64882f50290", "0xa324b88c18fb7109af79f03de4eee88bd6207376a085898c82d67b7d1a35ac41", 244]}' http://localhost:9933
```

Response (shortened):

```json
{
   "jsonrpc":"2.0",
   "result":"0x044e8e00c93905add716002173aa62feb3...83666edbb4da7ae3573c962f710f6e8500",
   "id":1
}
```

### 1.7.3. grandpa_subscribeJustifications (pubsub)

Returns the block most recently finalized by Grandpa, alongside side its justification.

This endpoints communicates over the Websocket protocol (`grandpa_justifications` subscription).

#### 1.7.3.1. Parameter

None.

#### 1.7.3.2. Response

* `HEX` - The block itself.

#### 1.7.3.3. Example

Request:

```json
{
    "id": 1,
    "jsonrpc": "2.0",
    "method": "grandpa_subscribeJustifications",
    "params": []
}

> ```

Author: I'm not sure about this. Curl examples are quite useful since people can just copy paste those in their own environment (which runs Substrate/Polkadot, for example). With websockets it's not that easy. Personally I used [wscat](https://github.com/websockets/wscat).

Looks like there is a `-x` parameter which can be used as a single command. I can adjust the examples if you think this should be added, because you're right about readability.

    "id": 1
}
```

### 1.8.7. author_hasSessionKeys

Checks if the keystore has private keys for the given session public keys.

#### 1.8.7.1. Parameter

* `HEX` - The SCALE encoded, concatenated keys.

#### 1.8.7.2. Response

* `BOOL` - Returns `true` if all private keys could be found, `false` if otherwise.

#### 1.8.7.3. Example

Request:

```bash
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "author_hasSessionKeys", "params":["0x8c0baa0b4cf42e669a3805c1d6405926c9adf0691f854a6ddaffde3abc1dbd6b7c70cc7d2a731186ec54e26e0b0509667376d818643a5969549e44a76dc42f5a041b0120d2fc6d686e1bec66f596ddcce78da4029a23b3c213c55d2a064e9c26a20ab338080835b845e71573c3197795b729a1015b504a8352ee7dcbce92296c28bcc393e3cb1b18f597c597a458e21f706374e75f35445132977a66519d731d"]}' http://localhost:9933
```

Response:

```json
{
    "jsonrpc": "2.0",
    "result": true,
    "id": 1
}
```

### 1.8.8. author_hasKey

Checks if the keystore has private keys for the given public key and key type.

#### 1.8.8.1. Parameter

* `HEX` - The public key.

> * `STRING` * [Key type](#Key-types).

Author: I have `*` pretty much everywhere for lists (primarily because my markdown plugin was complaining about `-`). Should I adjust this document-wide?

> * `STRING` * [Key type](#Key-types).

Author: Well, `prettier` took care of that part.

                "number": "0x12d2dd",
                "parentHash": "0x78b7af6942c220ec33945f228f1e17383403561f8d7bbc02da2c79650f22fdbe",
                "stateRoot": "0xa6387842acac9315ad4424828f571c348a40bdfb9f9b4cb4951bcb76e43dacfb"
            }
        },
        "justification": null
    },
    "id": 1
}
```

### 1.9.4. chain_getBlockHash

Get hash of the 'n-th' block in the canon chain. If no parameters are provided, the latest block hash gets returned.

Alias: `chain_getHead`

#### 1.9.4.1. Parameter

One of the following types is allowed:

* `HEX` - (OPTIONAL) The value indicating the 'n-th' block in the chain.
* `U32` - (OPTIONAL) The value indicating the 'n-th' block in the chain.
* `ARRAY` - (OPTIONAL) Multiple values, where each value can be one of the following types:
  * `HEX` - The value indicating the 'n-th' block in the chain.
  * `U32` - The value indicating the 'n-th' block in the chain.

#### 1.9.4.2. Response

One of the following types gets returned, depending on the provided parameters:

* `HASH` - (OPTIONAL) The block hash.
* `ARRAY`
  * `HASH` - (OPTIONAL) A block hash

#### 1.9.4.3. Example

Request:

```bash

> curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chain_getBlockHash", "params":[50, "0x64", [200]]}' http://localhost:9933

Reviewer 2: One of seem to suggest just one of them, here the example uses all three at once. Is that as intended?

Author: The parameters on this method work somewhat interestingly. You can specify one or multiple block numbers, either as integers or hex string, and can even include embedded arrays. But I guess this is not really "good design", so I'll adjust this (backwards compatible with substrate).

#### 1.11.4.1. Parameter

* `HEX` - The prefix.
* `HEX` - (OPTIONAL) The block hash.

#### 1.11.4.2. Response

* `ARRAY`
  * `ARRAY` - (OPTIONAL)
    * `HEX` - The key.
    * `HEX` - The value.

#### 1.11.4.3. Example

Request:

```bash
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "state_getPairs", "params":["0x", null]}' http://localhost:9933
```

Response (shortened):

```json
{
    "jsonrpc": "2.0",
    "result": [
        [
            "0x0b76934f4cc08dee01012d059e1b83eebbd108c4899964f707fdaffb82636065",
            "0x00"
        ],
        [
            "0x11f3ba2e1cdd6d62f2ff9b5589e7ff816254e9d55588784fa2a62b726696e2b1",
            "0x7b000000"
        ]
    ],
    "id": 1
}
```

### 1.11.5. state_getKeysPaged

**Warning**: This method is [UNSAFE](#Safety).

Returns the keys with prefix with pagination support.

#### 1.11.5.1. Parameter

* `HEX` - (OPTIONAL) The prefix.
* `U32` - Amount of keys to be return.
* `HEX` - (OPTIONAL) The storage key after which the next keys should be returned in lexicographic order.
* `HASH` - (OPTIONAL) The block hash.

#### 1.11.5.2. Response

* `ARRAY`

>   * `STRING` - (OPTIONAL) Storage key

Author: Done, also found other sections where this was the case.
